paddle/phi/kernels/funcs/algorithm.h

```diff
@@ -71,7 +71,7 @@ HOSTDEVICE inline size_t UpperBound(const T1 *x, size_t num, const T2 &val) {
   while (count > 0) {
     auto step = (count >> 1);
     auto *it = first + step;
-    if (val < *it) {
+    if (static_cast<float>(val) < static_cast<float>(*it)) {
```

Comment:
Suggest that do not cast to 'float' directly, please use the MyTypeTrait in the document to infer the data type you need cast to.